### ISSUE TITLE
Rancher server has poor performing regex's under certain conditions

### ISSUE THREAD

**Issue by jdonofrio728** [rancher/rancher#7641]
We occasionally notice the our rancher-server host is eating up all of its CPU. upon investigation, the Java process inside the container had a few threads with the same stack traces:

https://gist.github.com/jdonofrio728/4afab9935a4eae6352bbd9351f422e32

A heap dump showed that the regex being applied is:
^((?:(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])(?:(?:\Q.\E(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+)?(?:\Q:\E[0-9]+)?\Q/\E)?[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?(?:(?:\Q/\E[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?)(?:\Q:\E([\w][\w.-]{0,127}))?(?:\Q@\E([A-Za-z][A-Za-z0-9]*(?:[-_+.][A-Za-z][A-Za-z0-9]*)*[:][0-9A-Fa-f]{32,}))?$

and the string object that is being matched is:
docker-virtual.ngp-artifactory/react-boilerplate-ui:

This is causing us to have to kill and restart docker server everytime it gets in this state.

**Rancher Versions:**
Server: 1.3.1
healthcheck:v0.2.0
ipsec:holder
network-services:v0.4.0
scheduler:v0.5.1
kubernetes (if applicable):

**Docker Version:**
1.13

**OS and where are the hosts located? (cloud, bare metal, etc):**
RHEL 7.2 vmware virtual machines

**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
HA Rancher with internal DB, 4 environments each with 3 nodes.

**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)**
Cattle

**Steps to Reproduce:**
I'm not entirely sure how to reproduce on using Rancher as we have not tracked down which call to triggers this behavior. However I can easily reproduce this issue in a groovy script using the latest build of cattle-docker-common.jar file in my classpath:

```
import io.cattle.platform.docker.client.DockerImage

startTime = System.currentTimeMillis()
image = DockerImage.parse("docker-virtual.ngp-artifactory/react-boilerplate-ui:")
endTime = System.currentTimeMillis()
time = endTime - startTime
print "Time taken to parse: ${time}"
```

**Results:** 
The results of my mini test show that it takes an indefinite amount of time to parse the string using the coded regex. Simple docker image names parse quickly, however the string in our environment spins indefinitly.

**Comment by vincent99** [comment]
While clearly it shouldn't do this regardless I just want to clarify since this issue originally (from email) and now contain two different image IDs, both of which are invalid:

- `docker-virtual.ngp-artifactory/react-boilerplate-ui:`
-`docker-virtual.ngp-artifactory/react-boilerplate-ui::latest`

Is this happening with one of those, both, or something else?

**Comment by jdonofrio728** [comment]
My mistake, its happening with ```docker-virtual.ngp-artifactory/react-boilerplate-ui:```

**Comment by vincent99** [comment]
@StrongMonkey https://github.com/rancher/cattle/commit/2e0f83b475b1e227e634c9674a2488b1607ac722

**Comment by StrongMonkey** [comment]
@vincent99 that java class is strictly derived from a go package which Docker use to parse their image repo name. Not sure why it takes such a long time to parse an invalid name, i ll investigate more.

**Comment by vincent99** [comment]
Basically there's an exponential number of backtracks required to handle the `:`, `O(2^(length of input string))`.  The worst case is when the pattern doesn't actually match the input at all, which it doesn't here because ending in `:` with no tag is invalid.  And being longer than a typical image string makes it even worse.

http://www.regular-expressions.info/catastrophic.html

TL;DR: 

> Some people, when confronted with a problem, think "I know, I'll use regular expressions." Now they have two problems.
https://blog.codinghorror.com/regular-expressions-now-you-have-two-problems

**Comment by vincent99** [comment]
Note that the current version of Docker CLI catches this somehow and fails quickly:

```
docker run docker-virtual.ngp-artifactory/react-boilerplate-ui:
docker: Error parsing reference: "docker-virtual.ngp-artifactory/react-boilerplate-ui:" is not a valid repository/tag: invalid reference format.
See 'docker run --help'.
```

**Comment by cjellick** [comment]
We should investigate how docker impl'd, but I'm not against explicitly checking for and rejecting images that end in `:`

**Comment by StrongMonkey** [comment]
@cjellick i am not exactly using the same library as docker does. (the regex library differs from java to go). But i do notice the java code do have a bad time parsing this kind of docker images...

**Comment by StrongMonkey** [comment]
Can either be addressed by https://github.com/rancher/cattle/pull/2762 or https://github.com/rancher/cattle/pull/2761

**Comment by StrongMonkey** [comment]
Available at master now. Note that this change would completely drop the image validation in cattle.

**Comment by sangeethah** [comment]
Tested with rancher-server build from v1.6 branch.
There is no validation done on image field anymore.

When we try to start a service with invalid image (docker-virtual.ngp-artifactory/react-boilerplate-ui:/docker-virtual.ngp-artifactory/react-boilerplate-ui::latest), we get the following error message from the agent and container gets to "Error" state:

```Failed to allocate instance [container:1i133]: Bad instance [container:1i133] in state [erroring]: invalid reference format```

Before the fix , we would have failed at the the api level with error message - ```Validation failed in API: imageUuid ```
